futures-util/src/task/arc_wake.rs

```diff
@@ -32,39 +23,43 @@ pub trait ArcWake {
     ///
     /// If `wake()` is called on the returned `Waker`,
     /// the `wake()` function that is defined inside this trait will get called.
-    fn into_waker(wake: Arc<Self>) -> Waker where Self: Sized
+    fn into_waker(self: Arc<Self>) -> Waker where Self: Sized
     {
-        let ptr = Arc::into_raw(wake) as *const();
+        let ptr = Arc::into_raw(self) as *const();
 
         unsafe {
             Waker::new_unchecked(RawWaker::new(ptr, waker_vtable!(Self)))
         }
     }
 }
 
+// FIXME: panics on Arc::clone / refcount changes could wreak havoc on the
+// code here. We should guard against this by aborting.
+
 unsafe fn increase_refcount<T: ArcWake>(data: *const()) {
     // Retain Arc by creating a copy
     let arc: Arc<T> = Arc::from_raw(data as *const T);
     let arc_clone = arc.clone();
     // Forget the Arcs again, so that the refcount isn't decrased
-    let _ = Arc::into_raw(arc);
-    let _ = Arc::into_raw(arc_clone);
+    mem::forget(arc);
+    mem::forget(arc_clone);
 }
 
-unsafe fn clone_arc_raw<T: ArcWake>(data: *const()) -> RawWaker {
+// used by `waker_ref`
+pub(super) unsafe fn clone_arc_raw<T: ArcWake>(data: *const()) -> RawWaker {
```

Comment:
Where are those things exported to now? Only one level above, or even further?
It's totally to fine to use those within other parts of futures-util, but I don't think the library should publicly export these.

Reply:
They're just `pub(super)`-- so only available to the futures_util::task module.